server/src/main/java/io/deephaven/server/object/ObjectServiceGrpcImpl.java

```diff
@@ -0,0 +1,196 @@
+        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+            final SessionState session = sessionService.getCurrentSession();
+            final String type = request.getSourceId().getType();
+            if (type.isEmpty()) {
+                throw GrpcUtil.statusRuntimeException(Code.FAILED_PRECONDITION, "No type supplied");
+            }
+            if (request.getSourceId().getTicket().getTicket().isEmpty()) {
+                throw GrpcUtil.statusRuntimeException(Code.FAILED_PRECONDITION, "No ticket supplied");
+            }
+            final SessionState.ExportObject<Object> object = ticketRouter.resolve(
+                    session, request.getSourceId().getTicket(), "sourceId");
+            session.nonExport()
+                    .require(object)
+                    .onError(responseObserver)
+                    .submit(() -> {
+                        final Object o = object.get();
+                        final FetchObjectResponse response = serialize(type, session, o);
+                        responseObserver.onNext(response);
+                        responseObserver.onCompleted();
+                        return null;
+                    });
+        });
+    }
+
+    private FetchObjectResponse serialize(String expectedType, SessionState state, Object object) throws IOException {
+        final ExposedByteArrayOutputStream out = new ExposedByteArrayOutputStream();
+        final ObjectType objectType =
+                objectTypeLookup.findObjectType(object).orElseThrow(() -> noTypeException(object));
+        if (!expectedType.equals(objectType.name())) {
+            throw new IllegalArgumentException(String.format("The expected type '%s' is not equal to found type '%s'",
+                    expectedType, objectType.name()));
+        }
+        final ExportCollector exportCollector = new ExportCollector(state);
+        try {
+            objectType.writeTo(exportCollector, object, out);
+            final Builder builder = FetchObjectResponse.newBuilder()
+                    .setType(objectType.name())
+                    .setData(ByteStringAccess.wrap(out.peekBuffer(), 0, out.size()));
+            for (ReferenceImpl ref : exportCollector.refs()) {
+                builder.addTypedExportId(ref.typedTicket());
+            }
+            return builder.build();
+        } catch (Throwable t) {
+            cleanup(exportCollector, t);
+            throw t;
+        }
+    }
+
+    private static void cleanup(ExportCollector exportCollector, Throwable t) {
+        for (ReferenceImpl ref : exportCollector.refs()) {
+            try {
+                ref.export.release();
+            } catch (Throwable inner) {
+                t.addSuppressed(inner);
+            }
+        }
+    }
+
+    private static IllegalArgumentException noTypeException(Object o) {
+        return new IllegalArgumentException(
+                "No type registered for object class=" + o.getClass().getName() + ", value=" + o);
```

Comment:
possibly should be a failed precondition instead, so the user gets feedback? I probably could be convinced either way on this...

Only risk I see with sending a response "sorry that can't be serialized" is that we end up giving some detail about the actual object. Omitting the class name from the public message and logging it would solve that for me?